### PR TITLE
feat: add NESTED_COMMIT markers when combining multiple source commits

### LINKED ISSUE #3342: Owlbot multiple commit descriptions can miss release notes
This commit https://github.com/googleapis/nodejs-spanner/commit/d1a9d7c3da7c378b433dcdab459b0c46279a4fb2 has "fix: annotating some fields as REQUIRED" in the description. Apparently this commit has multiple "PiperOrigin-RevId:" associated with it and the first entry is "chore: update v2.14.1 gapic-generator-typescript".

Screen Shot 2022-03-23 at 1 00 22 PM

The commit was released as part of [v5.17.0](https://github.com/googleapis/nodejs-spanner/releases/tag/v5.17.0)

Screen Shot 2022-03-23 at 1 01 00 PM

However, the v5.17.0 release note https://github.com/googleapis/nodejs-spanner/releases/tag/v5.17.0 does not have the entry "fix: annotating some fields as REQUIRED".

Screen Shot 2022-03-23 at 1 03 39 PM

Would you troubleshoot and fix the problem?

CC: @bcoe

### PR DESCRIPTION
**Pull request by chingor13**
In https://github.com/googleapis/release-please/pull/1566, we added the ability to explicitly denote multiple commit messages by using `BEGIN_NESTED_COMMIT`/`END_NESTED_COMMIT` markers.

This PR now augments owl-bot to include those markers when combining multiple copy-code runs in a single PR. We add a new `prependCommitMessage` helper that encapsulates the logic of recombining a previous owl-bot copy PR, wrapping the previous message in a `BEGIN/END_NESTED_COMMIT` block and pre-pending the new commit message.

Note that any existing PR bodies will be grouped together into a single nested commit when the next copy code commit is added.

Fixes #3342

### CODE REVIEW

**Comment by chingor13** [comment]
Blocked until #4427 is deployed

File: packages/owl-bot/__snapshots__/create-pr.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+exports[
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/__snapshots__/create-pr.js:1]: I've never seen a __snapshots__ directory before.  What is it for?

File: packages/owl-bot/src/create-pr.ts [hunk 2]
```diff
@@ -53,6 +58,56 @@ export function resplit(
   return {title, body: body.substring(0, MAX_BODY_LENGTH)};
 }
 
+const NESTED_COMMIT_SEPARATOR = 'BEGIN_NESTED_COMMIT';
+
+/**
+ * Given pull request content and a new commit message. Rewrite the pull request
+ * title and body. If the initial pull request title was truncated with ellipses,
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/create-pr.ts:65]: Rewrite it how?  Please add sample input and output to the comment.

File: packages/owl-bot/src/create-pr.ts [hunk 3]
```diff
@@ -53,6 +58,56 @@ export function resplit(
...
+      : // there are already existing nested commit tags, content before the first
+        // one is wrapped in a new nested commit tag
+        `${NESTED_COMMIT_SEPARATOR}\n${
+          bodyParts[0]
+        }\nEND_NESTED_COMMIT\n${bodyParts
+          .slice(1)
+          .join(NESTED_COMMIT_SEPARATOR)}`;
+
+  // prepend the new commit message and use original title truncation logic
+  return resplit(
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/create-pr.ts:105]: Why not call prependCommitMessage before calling resplit?  Then resplit gets called only once and the unsplitting logic in this function can be removed.

File: packages/owl-bot/test/create-pr.ts [hunk 4]
```diff
@@ -97,6 +99,40 @@ describe('resplit', () => {
...
+        const pullContent = resplit(
+          'feat: some feature\n\nadditional context',
+          withCheckbox
+        );
+        const prependedContent = prependCommitMessage(
+          'fix: some new feature\n\nmore additional context',
+          pullContent,
+          withCheckbox
+        );
+        snapshot(prependedContent.title);
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/test/create-pr.ts:118]: snapshot is hard to follow.  I'm much rather see two strings compared in line.

**Review by SurferJeffAtGoogle — COMMENTED** [review]
Please add a flag to enable this new behavior.

Do the delimiters get added when a PR contains only a single commit?

File: packages/owl-bot/test/create-pr.ts (hunk 4 above)
Comment (chingor13) [reply packages/owl-bot/test/create-pr.ts:118]: The snapshots are easier to construct code-wise at the expense of the asserted value living in a different file. It is useful for large text bodies that are harder to read inline. I can switch this back to inline assertions though.

**Comment by chingor13** [comment]
@SurferJeffAtGoogle I'm unsure how we want to opt-in this functionality. In code, I added an enum that is surfaced as a CLI argument, but I'm unsure how we want to configure an opt-in for a repository.

File: packages/owl-bot/test/create-pr.ts (hunk 4 above)
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/test/create-pr.ts]: Thank you for offering to switch back to inline assertions.  Please do.

File: packages/owl-bot/src/create-pr.ts (hunk 3 above)
Comment (SurferJeffAtGoogle) [reply packages/owl-bot/src/create-pr.ts]: > We do want to rejoin that previous message to be nicer for the release notes.

Indeed, so why not call this function on the original message (before resplit)?

**Comment by SurferJeffAtGoogle** [comment]
> I'm unsure how we want to opt-in this functionality. In code, I added an enum that is surfaced as a CLI argument, but I'm unsure how we want to configure an opt-in for a repository.

The CLI argument looks great!

**Comment by SurferJeffAtGoogle** [comment]
Do the delimiters get added when a PR contains only a single commit?

**Comment by chingor13** [comment]
> Do the delimiters get added when a PR contains only a single commit?

No, they are only added when we append to an existing PR

File: packages/owl-bot/src/copy-code.ts [hunk 5]
```diff
@@ -372,14 +373,11 @@ async function findAndAppendPullRequest(
...
-    `${commitBody}\n\n` +
-      `${pull.title}\n` +
-      pull.body
-        ?.replace(REGENERATE_CHECKBOX_TEXT, '')
-        .replace(EMPTY_REGENERATE_CHECKBOX_TEXT, '')
-        .trim() ?? '',
-    WithRegenerateCheckbox.Yes
+  const {title, body} = prependCommitMessage(
+    commitBody,
+    {title: pull.title, body: pull.body || ''},
```
Comment (SurferJeffAtGoogle) [inline packages/owl-bot/src/copy-code.ts:378]: Nit: the typescript idiom is `pull.body ?? ''`.

**Review by SurferJeffAtGoogle — APPROVED** [review]

**Comment by meltsufin** [comment]
@SurferJeffAtGoogle @chingor13 This still doesn't seem to be working on the Java monorepo. Do we need to add any config?
Example: https://github.com/googleapis/google-cloud-java/pull/8614
